src/manifold/src/manifold.cpp

```diff
@@ -138,6 +138,29 @@ Mesh Manifold::GetMesh() const {
   return result;
 }
 
+size_t Manifold::GetMeshDirect() const {
+  const Impl& impl = *GetCsgLeafNode().GetImpl();
+
+  int recs = 6 + (NumTri() * 3);
+  size_t* u = (size_t*)malloc(sizeof(size_t) * recs);
+  float* f = (float*)u;
+
+  int rp = 0;
+  // header
+  u[rp++] = NumVert();
+  u[rp++] = NumTri();
+  u[rp++] = (size_t)impl.vertPos_.begin();
+  u[rp++] = (size_t)impl.vertPos_.end();
+  u[rp++] = (size_t)impl.vertNormal_.begin();
+  u[rp++] = (size_t)impl.vertNormal_.end();
```

Comment:
I don't think we can safely pass around pointers to our internal data structures. First, they need to be pointers to const or someone could reek all kinds of havoc. Second, what if this impl goes out of scope before this pointer gets read? 

We'll need to copy the vert buffers just like the triangle buffer, but at least we can copy them into the right shape:
```
struct MeshGL {
  int numVert, numTri;
  std::unique_ptr<float[]> vertPos, vertNormal;
  std::unique_ptr<uint32_t[]> triVerts;
};

MeshGL out;
out.vertPos = std::make_unique<float[]>(3*numVert);
for(int i = 0; i < numVert; ++i) {
  glm::vec3 v = impl.vertPos_[i];
  out.vertPos[3*i] = v.x;
  out.vertPos[3*i + 1] = v.y;
  out.vertPos[3*i + 2] = v.z;
} 
```
And use `out.vertPos.get()` for the pointer in your `subarray` call in JS. Embind [supports](https://emscripten.org/docs/porting/connecting_cpp_and_javascript/embind.html#smart-pointers) `std::unique_ptr` out of the box.

Reply:
the JS + WASM VMs run in the same thread, so there is no chance the heap changes under you before the data is copied out. at least, this is true for the way the code is currently written. I'll try the approach above and see how that performs. I understand the desire for parity across targets. some performance gains won't necessarily translate.

Reply:
[trying and failing](https://github.com/stewartoallen/manifold/tree/direct-mesh-data-access-proper) to do this the right way ... not understanding how to complete the abstraction / binding / wrapping

Reply:
Agreed this would probably be safe for the WASM API, but our C++ API is even more important (that's the only one you can use currently to get GPU speedups). And yes, the Embind stuff is a bit convoluted. We're all figuring it out together - go ahead and push your attempt so we can look it over and give pointers.